subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -3263,11 +3268,13 @@ int lwm2m_discover_handler(struct lwm2m_message *msg, bool is_bootstrap)
 		}
 
 		/* For bootstrap discover, only report object ID when no
-		 * instance is available.
+		 * instance is available or it's needed to report object
+		 * version.
 		 * For device management discovery, only report object ID with
 		 * attributes if object ID (alone) was provided.
 		 */
-		if ((is_bootstrap && obj->instance_count == 0U) ||
+		if ((is_bootstrap && (obj->instance_count == 0U ||
+				      lwm2m_engine_report_obj_version(obj))) ||
```

Comment:
This function name suggests more to me that it will actively report the obj version rather than evaluate if it needs to be reported. If not for the comment above I would most certainly misunderstand this if statement

Reply:
Indeed on second thought, it could be misleading. How about `lwm2m_engine_shall_report_obj_version`?  A little lengthy, but there should be no confusion then.

Reply:
I've pushed the updated version.

Reply:
If I am to nitpick I'd say I prefer should over shall . But either way it is much better :)

Reply:
> If I am to nitpick I'd say I prefer should over shall .

+1. But mostly as my 2 cents, don't want to bother the maintainer with updating this long patch ;-).